Allow origin api controllers to be run independently in the master

smarterclayton opened the pull request:
First step to HA.

Run `openshift start master api --config=...` to run just the apiserver.  Run `openshift start master controllers --config=...` to run just the controllers.

Known gaps:
* [ ] no health check or metrics available, need to abstract the code for that
* [ ] no test cases yet

smarterclayton commented:
@sdodson fyi, @deads2k or @liggitt or @stevekuznetsov review

smarterclayton commented:
@ironcladlou so you don't have to jump through hoops:

    openshift start --write-config=config1
    openshift start master --config=config1/master/master-config.yaml
    openshift start master controllers --config=config1/master/master-config.yaml

Two controller loops, one apiserver.

smarterclayton commented:
@timothysc fyi this is closer to what we discussed.  we'll do the leader lease next.  I'm not going to break controllers up yet, but this gives you two independent pieces

timothysc commented:
@rrati - lets look over in the a.m.

timothysc commented:
@sdodson @detiber @brenton @rrati - We should chat as this would change deployment.  Ideally you would want a unit file for each mode.

api 
everything else

This would effect pacemaker config as well.

stevekuznetsov commented:
@smarterclayton run `hack/update-generated-completions.sh`

deads2k commented on pkg/cmd/server/start/start_allinone.go:
```diff
@@ -20,22 +20,18 @@ import (
-	PrintIP          bool
-	Output           io.Writer
-	DisabledFeatures []string
+	MasterOptions
```
I really dislike anonymous inclusion because of the unexpected interface fulfillment and name hiding.  Do you object to giving this a name?

smarterclayton replied:
No objection

On Aug 13, 2015, at 9:02 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/server/start/start_allinone.go
<https://github.com/openshift/origin/pull/4132#discussion_r36970911>:

>  	"github.com/openshift/origin/pkg/cmd/server/start/kubernetes"
>  	cmdutil "github.com/openshift/origin/pkg/cmd/util"
>  )
>
>  type AllInOneOptions struct {
> -	MasterArgs *MasterArgs
> -	NodeArgs   *NodeArgs
> -
> -	CreateCerts      bool
> -	ConfigDir        util.StringFlag
> -	MasterConfigFile string
> -	NodeConfigFile   string
> -	PrintIP          bool
> -	Output           io.Writer
> -	DisabledFeatures []string
> +	MasterOptions

I really dislike anonymous inclusion because of the unexpected interface
fulfillment and name hiding. Do you object to giving this a name?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4132/files#r36970911>.

deads2k commented:
Vision question.  If we get HA and active/active working correctly, would we ever want to suggest a configuration other than two full masters (complete with API server and controllers) running together?  That would seem simplify our topology story at the expense of scaling out the API server and controllers independently.  Unless we expect a significant performance degradation due to controller interactions, I'd expect that we'd want to take the simpler approach.

If we'd rather scale out a single thing, then it seems to make sense to support a `json` format config (or use some bash transformations) to modify a config in a bash script using `jq` for our extended tests.

deads2k commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,76 @@
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+const apiLong = `Start the master controllers
```
bad description

timothysc commented:
@deads2k Our plan has always been to go active-active on apiserver and still maintain active-passive on other components, but they become "warmer".  active-active on the api will yield the end user perspective that the system is always running.

However, the extra load that is taken on by existing upstream controllers is cost-prohibitive to go active-active (kcm+sched) because they effect performance with little gain unless the system is in flux (scheduling in progress, updating replicas, etc.) Each additional openshift component of would need to be looked at separately.

deads2k commented:
> @deads2k Our plan has always been to go active-active on apiserver and still maintain active-passive on other components, but they become "warmer". active-active on the api will yield the end user perspective that the system is always running.

The question is more about how you see a customer configuring this.  Would they be responsible to individually starting api servers and controllers?  If not, this change isn't moving the direction we want.  If so, it's fine.  I think that even an active/passive setup is better served by electing the leader automatically and having a single entity to start up.

timothysc commented:
The complexity would have to be in the installer... right?

smarterclayton commented:
Yeah, three paths:

1. Installer, split
2. Running on infra, split
3. Simple setup and dev, combined.

In the long run, a great installer takes this off the user's plate in a way
that has little downside for us. In the short term, trying to be careful.

Biggest advantage of split is that losing the leader lease lets you shoot
yourself in the head with little impact to users, which means we can keep
API proxy connections open longer.

On Aug 13, 2015, at 10:19 AM, Timothy St. Clair <notifications@github.com>
wrote:

The complexity would have to be in the installer... right?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4132#issuecomment-130693025>.

deads2k commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,76 @@
+	options.MasterArgs.StartControllers = false
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
```
Do we want to require a full `MasterConfig`?  Not all of the information is required just to start the API server.

Also, there are likely to be commands related to masters like, "validate-config", and "write-config".  Would we prefer `openshift master start|[config write|validate|migrate|view-with-defaulting]`?

smarterclayton replied:
Down the road, maybe.  For now, I'd like to keep it simple.

On Thu, Aug 13, 2015 at 10:47 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r36982584>:
>
> > +						for _, cause := range details.Causes {
> > +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
> > +						}
> > +						os.Exit(255)
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartControllers = false
> > +
> > +	flags := cmd.Flags()
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
>
> Do we want to require a full MasterConfig? Not all of the information is
> required just to start the API server.
>
> Also, there are likely to be commands related to masters like,
> "validate-config", and "write-config". Would we prefer openshift master
> start|[config write|validate|migrate|view-with-defaulting]?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r36982584>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton replied:
Comments addressed.

On Thu, Aug 13, 2015 at 11:13 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Down the road, maybe.  For now, I'd like to keep it simple.
>
> On Thu, Aug 13, 2015 at 10:47 AM, David Eads <notifications@github.com>
> wrote:
>
>> In pkg/cmd/server/start/start_api.go
>> <https://github.com/openshift/origin/pull/4132#discussion_r36982584>:
>>
>> > +						for _, cause := range details.Causes {
>> > +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
>> > +						}
>> > +						os.Exit(255)
>> > +					}
>> > +				}
>> > +				glog.Fatal(err)
>> > +			}
>> > +		},
>> > +	}
>> > +
>> > +	options.MasterArgs = NewDefaultMasterArgs()
>> > +	options.MasterArgs.StartControllers = false
>> > +
>> > +	flags := cmd.Flags()
>> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
>>
>> Do we want to require a full MasterConfig? Not all of the information is
>> required just to start the API server.
>>
>> Also, there are likely to be commands related to masters like,
>> "validate-config", and "write-config". Would we prefer openshift master
>> start|[config write|validate|migrate|view-with-defaulting]?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/4132/files#r36982584>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

deads2k commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,76 @@
+will run in the foreground until you terminate the process.`
+
+// NewCommandStartMasterAPI starts only the APIserver
+func NewCommandStartMasterAPI(fullName string, out io.Writer) (*cobra.Command, *MasterOptions) {
```
`name, fullName string`

deads2k commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,76 @@
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
Do we care about the "launching without config" aspect of this command?  It doesn't seem like one we'd be likely to just decide to use.

smarterclayton replied:
I thought about it and decided no.  We *want* people to use config.

On Aug 13, 2015, at 10:50 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/server/start/start_api.go
<https://github.com/openshift/origin/pull/4132#discussion_r36982931>:

> +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
> +						}
> +						os.Exit(255)
> +					}
> +				}
> +				glog.Fatal(err)
> +			}
> +		},
> +	}
> +
> +	options.MasterArgs = NewDefaultMasterArgs()
> +	options.MasterArgs.StartControllers = false
> +
> +	flags := cmd.Flags()
> +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> +	BindListenArg(options.MasterArgs.ListenArg, flags, "")

Do we care about the "launching without config" aspect of this command? It
doesn't seem like one we'd be likely to just decide to use.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4132/files#r36982931>.

deads2k replied:
> I thought about it and decided no.  We *want* people to use config.

In that case, the `ListenArg` should be extraneous.

smarterclayton replied:
See below

deads2k commented on pkg/cmd/server/start/start_allinone.go:
```diff
@@ -61,12 +57,8 @@ You may also pass --kubeconfig=<path> to connect to an external Kubernetes clust
 
 // NewCommandStartAllInOne provides a CLI handler for 'start' command
 func NewCommandStartAllInOne(fullName string, out io.Writer) (*cobra.Command, *AllInOneOptions) {
```
Renamed `fullName`.  It isn't a full name in the same sense as our other commands and now that it has a semantic meaning (used below in `DefaultsFromName`).

deads2k commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,76 @@
+will run in the foreground until you terminate the process.`
+
+// NewCommandStartMasterAPI starts only the APIserver
+func NewCommandStartMasterAPI(name, fullName string, out io.Writer) (*cobra.Command, *MasterOptions) {
```
Rename `fullName` to reflect what it is.  `baseName`?

deads2k commented on pkg/cmd/server/start/master_args.go:
```diff
@@ -44,6 +44,8 @@ type MasterArgs struct {
 	MasterPublicAddr flagtypes.Addr
 
+	StartAPI         bool
+	StartControllers bool
```
Suggestion: Common usage of `NewDefaultMasterArgs` followed by setting the `Start*` bits individually to `false` requires knowledge of all the flags.  Having `NewDefaultMasterArgs` return with the `Start*` bits disabled, will keep things like `openshift start master api` operating as intended if we add more switches.

deads2k commented on pkg/cmd/server/start/start_controllers.go:
```diff
@@ -0,0 +1,76 @@
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
I don't think this is required when starting from config.

smarterclayton replied:
It is, because we need an address to listen on for health check and metrics.  It's unused now but that's in the checklist.  It can't be the value from master because you might run controllers on the same host and you can't bind to the same name.

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -265,11 +265,24 @@ func (o MasterOptions) RunMaster() error {
+		if err != nil {
+			return err
+		}
+		return StartControllers(openshiftConfig, kubeConfig)
```
I don't like the way this logic is built.  Why not simply build the `masterConfig` with correct values and have `StartMaster(*origin.MasterConfig)` start the correct pieces as it goes?  That seems more maintainable if we're going to use this as the common entry point.

deads2k replied:
If we keep the separate `StartControllers`, I think I'm inclined to also make a separate `StartAPIServer` and call them directly from the commands.

smarterclayton replied:
Maybe in the future.  But now, it's not worth the larger refactor.  I want to wait until we decide what to do with the UI and the oauth server

smarterclayton replied:
Addressed naming and defaulting behavior for StartAPI

On Thu, Aug 13, 2015 at 2:55 PM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_master.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37011600>:
>
> >  	}
> >
> > -	return nil
> > +	if !o.MasterArgs.StartAPI {
> > +		openshiftConfig, err := origin.BuildMasterConfig(*masterConfig)
> > +		if err != nil {
> > +			return err
> > +		}
> > +
> > +		_, kubeConfig, err := buildKubernetesMasterConfig(openshiftConfig)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		return StartControllers(openshiftConfig, kubeConfig)
>
> If we keep the separate StartControllers, I think I'm inclined to also
> make a separate StartAPIServer and call them directly from the commands.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37011600>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

deads2k replied:
> Maybe in the future. But now, it's not worth the larger refactor. I want to wait until we decide what to do with the UI and the oauth server

Can you add a comment saying this behavior shouldn't be extended?  I'm having flashbacks to the old start paths that took over a week to unravel.

smarterclayton commented:
[test]

deads2k commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,76 @@
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
I think you'll want comments explaining why these are here or someone will pull them out when they notice.

liggitt replied:
I also get twitchy about mixing config and flags... definitely want a comment explaining why this exists, and a TODO to improve it

smarterclayton replied:
Comments addressed

On Fri, Aug 14, 2015 at 12:39 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37094201>:
>
> > +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
> > +						}
> > +						os.Exit(255)
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartAPI = true
> > +
> > +	flags := cmd.Flags()
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> > +	BindListenArg(options.MasterArgs.ListenArg, flags, "")
>
> I also get twitchy about mixing config and flags... definitely want a
> comment explaining why this exists, and a TODO to improve it
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37094201>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

deads2k commented:
I'm a little concerned about fragility moving forward without refactoring the start path, but I suppose a couple comments explaining why (and issues open to fix) potentially conflicting flag bindings are present, why the launch path runs on reverse logic that doesn't follow the flags directly, and about why additional bool flags need to rippled through unrelated code would keep me from whining too badly.

I'm assuming this is time critical enough that you aren't going to fix those problems?

smarterclayton commented:
I don't think we've made the code worse.  Having another split will make
the refactor clearer.  I'll add the comments.

What do you mean by reverse logic?

On Fri, Aug 14, 2015 at 8:34 AM, David Eads <notifications@github.com>
wrote:

> I'm a little concerned about fragility moving forward without refactoring
> the start path, but I suppose a couple comments explaining why (and issues
> open to fix) potentially conflicting flag bindings are present, why the
> launch path runs on reverse logic that doesn't follow the flags directly,
> and about why additional bool flags need to rippled through unrelated code
> would keep me from whining too badly.
>
> I'm assuming this is time critical enough that you aren't going to fix
> those problems?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132#issuecomment-131093632>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

deads2k commented:
> What do you mean by reverse logic?

For the two commands you added, I think the logic would read easier with refactor to allow reading the config and only starting the pieces of the master you want directly in the command.  The current implementation has:
 1. a flag declared in the positive direction: https://github.com/openshift/origin/pull/4132/files#diff-36c6d5003e486e2b6479632137743725R48
 1. a test against the reverse (not): https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR270
 1. a second flag set in the opposing direction: https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR271
 1. another notted check: https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR274
 1. that makes a call a method you'd expect to be called by the other flag: https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR284

It's hard to parse what happens and even then, weird stuff happens.   Consider the "both false" case, where we set a flag, then actually call a startControllers method that will no-op.  

I think that flow is worth refactoring, but I do understand internal beauty for time considerations.

liggitt commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -314,6 +329,14 @@ func (o MasterOptions) CreateCerts() error {
+	if openshiftConfig.Options.KubernetesMasterConfig == nil {
+		return false, nil, nil
+	}
+	kubeConfig, err := kubernetes.BuildKubernetesMasterConfig(openshiftConfig.Options, openshiftConfig.RequestContextMapper, openshiftConfig.KubeClient())
```
call this something other than `kubeConfig` or I will lose my mind

liggitt commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -371,25 +394,20 @@ func StartMaster(openshiftMasterConfig *configapi.MasterConfig) error {
-			return err
-		}
-
+	startKubeMaster, kubeConfig, err := buildKubernetesMasterConfig(openshiftConfig)
```
`kubeMasterConfig`

liggitt commented on pkg/cmd/server/start/start_api.go:
```diff
@@ -0,0 +1,77 @@
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	cmd.MarkFlagFilename("config", "yaml", "yml")
+	// TODO: add health and metrics endpoints on this listen argument
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
can we at least let the API use the configured bind address, and just let the controllers command do the weird config+arg thing? controllers would *just* expose health/metrics... if you change the API port here, you probably invalidate a lot of other bits of the config

smarterclayton replied:
I'm confused what you're asking.

On Fri, Aug 14, 2015 at 1:26 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37098424>:
>
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartAPI = true
> > +
> > +	flags := cmd.Flags()
> > +	// This command only supports reading from config and the listen argument
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> > +	cmd.MarkFlagFilename("config", "yaml", "yml")
> > +	// TODO: add health and metrics endpoints on this listen argument
> > +	BindListenArg(options.MasterArgs.ListenArg, flags, "")
>
> can we at least let the API use the configured bind address, and just let
> the controllers command do the weird config+arg thing? controllers would
> *just* expose health/metrics... if you change the API port here, you
> probably invalidate a lot of other bits of the config
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37098424>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

deads2k replied:
> I'm confused what you're asking.

The API server should use the listen port specified in the config file.  The controllers have a reasonable need to respecify, but I'm hard pressed to think of a reason why the API server shouldn't simply use the same port it's using for serving the API.

smarterclayton replied:
It should, see the updates where I removed it.

On Mon, Aug 17, 2015 at 9:56 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37189372>:
>
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartAPI = true
> > +
> > +	flags := cmd.Flags()
> > +	// This command only supports reading from config and the listen argument
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> > +	cmd.MarkFlagFilename("config", "yaml", "yml")
> > +	// TODO: add health and metrics endpoints on this listen argument
> > +	BindListenArg(options.MasterArgs.ListenArg, flags, "")
>
> I'm confused what you're asking.
>
> The API server should use the listen port specified in the config file.
> The controllers have a reasonable need to respecify, but I'm hard pressed
> to think of a reason why the API server shouldn't simply use the same port
> it's using for serving the API.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37189372>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
[merge] and will take more review comments in the follow up issues I'll spawn for listening for health and testing.

smarterclayton commented:
[merge]
